app/browser/menu.js

```diff
@@ -572,81 +578,72 @@ const createMenu = (CommonMenu) => {
     })
   }
 
-  const oldMenu = appMenu
+  let oldMenu = appMenu
   appMenu = Menu.buildFromTemplate(template)
   Menu.setApplicationMenu(appMenu)
   if (oldMenu) {
     oldMenu.destroy()
   }
 }
 
-const updateMenu = (CommonMenu, appState, windowData) => {
-  const updated = menuUtil.checkForUpdate(appState, windowData)
-  if (updated.nothingUpdated) {
-    return
-  }
-
-  // When bookmarks are removed via AppStore (context menu, etc), `isBookmarkChecked` needs to be recalculated
-  if (windowData && windowData.get('location')) {
-    isBookmarkChecked = isSiteBookmarked(appState.get('sites'), Immutable.fromJS({location: windowData.get('location')}))
-  }
-
-  // Only rebuild menus when necessary
-
-  if (updated.settings || updated.closedFrames) {
-    let historyMenu = menuUtil.getParentMenuDetails(appMenu, locale.translation('history'))
-    if (historyMenu && historyMenu.menu && historyMenu.menu.submenu && historyMenu.index !== -1) {
-      const menu = historyMenu.menu.submenu
-      const menuItems = createHistorySubmenu(CommonMenu)
-      menu.clear()
-      menuItems.forEach((item) => menu.append(new MenuItem(item)))
-    }
-  }
-
-  if (updated.sites) {
-    let bookmarksMenu = menuUtil.getParentMenuDetails(appMenu, locale.translation('bookmarks'))
-    if (bookmarksMenu && bookmarksMenu.menu && bookmarksMenu.menu.submenu && bookmarksMenu.index !== -1) {
-      const menu = bookmarksMenu.menu.submenu
-      const menuItems = createBookmarksSubmenu(CommonMenu)
-      menu.clear()
-      menuItems.forEach((item) => menu.append(new MenuItem(item)))
-    }
+const doAction = (action) => {
+  switch (action.actionType) {
+    case windowConstants.WINDOW_SET_FOCUSED_FRAME:
+      // TODO check/uncheck menu item instead of recreating menu
+      currentLocation = action.frameProps.get('location')
+      let menuItem = menuUtil.getMenuItem(appMenu, locale.translation('bookmarkPage'))
+      menuItem.checked = isCurrentLocationBookmarked()
+      break
+    case windowConstants.WINDOW_UNDO_CLOSED_FRAME:
+      appDispatcher.waitFor([appStore.dispatchToken], () => {
```

Comment:
[I had to lookup waitFor](https://facebook.github.io/flux/docs/dispatcher.html) and I think I understand this now

To make sure: this code is basically registering another handler for these actions **on top of the existing ones in windowStore**, right? This is **really** cool :smile: Great job!